internal/tiltfile/tiltfile_state.go

```diff
@@ -1110,6 +1118,67 @@ func (s *tiltfileState) translateK8s(resources []*k8sResource, updateSettings mo
 	return result, nil
 }
 
+func (s *tiltfileState) k8sDeployTarget(targetName model.TargetName, r *k8sResource, imageTargets []model.ImageTarget, updateSettings model.UpdateSettings) (model.K8sTarget, error) {
+	var kdTemplateSpec *v1alpha1.KubernetesDiscoveryTemplateSpec
+	if len(r.extraPodSelectors) != 0 {
+		kdTemplateSpec = &v1alpha1.KubernetesDiscoveryTemplateSpec{
+			ExtraSelectors: k8s.SetsAsLabelSelectors(r.extraPodSelectors),
+		}
+	}
+
+	sinceTime := apis.NewTime(pkgInitTime)
+	applySpec := v1alpha1.KubernetesApplySpec{
+		Timeout:                         metav1.Duration{Duration: updateSettings.K8sUpsertTimeout()},
+		PortForwardTemplateSpec:         k8s.PortForwardTemplateSpec(s.defaultedPortForwards(r.portForwards)),
+		DiscoveryStrategy:               r.discoveryStrategy,
+		KubernetesDiscoveryTemplateSpec: kdTemplateSpec,
+		PodLogStreamTemplateSpec: &v1alpha1.PodLogStreamTemplateSpec{
+			SinceTime: &sinceTime,
+			IgnoreContainers: []string{
+				string(container.IstioInitContainerName),
+				string(container.IstioSidecarContainerName),
+			},
+		},
+	}
+
+	var deps []string
+	if r.customDeploy != nil {
+		deps = r.customDeploy.deps
+		applySpec.Cmd = &v1alpha1.KubernetesApplyCmd{
+			Args: r.customDeploy.cmd.Argv,
+			Dir:  r.customDeploy.cmd.Dir,
+			Env:  r.customDeploy.cmd.Env,
+		}
+		applySpec.RestartOn = &v1alpha1.RestartOnSpec{
```

Comment:
fwiw - the pattern that i've found most helpful is to generate references to other API objects in the tiltfile reconciler, rather than in the tiltfile executor code. like so: https://github.com/tilt-dev/tilt/blob/master/internal/controllers/core/tiltfile/api.go#L56

mostly because it's easier to test that way.

Reply:
Ahhh interesting, I'll have to play around with that approach